src/libraries/System.Private.CoreLib/src/System/DateOnly.cs

```diff
@@ -0,0 +1,904 @@
+        /// <returns>An object that is equivalent to the date contained in s, as specified by format.</returns>
+        public static DateOnly ParseExact(ReadOnlySpan<char> s, ReadOnlySpan<char> format) => ParseExact(s, format, DateTimeFormatInfo.CurrentInfo, DateTimeStyles.None);
+
+        private const string OFormat = "yyyy'-'MM'-'dd";
+        private const string RFormat = "ddd, dd MMM yyyy";
+
+        /// <summary>
+        /// Converts the specified span representation of a date to its DateOnly equivalent using the specified format, culture-specific format information, and style.
+        /// The format of the string representation must match the specified format exactly or an exception is thrown.
+        /// </summary>
+        /// <param name="s">A span containing the characters that represent a date to convert.</param>
+        /// <param name="format">A span containing the characters that represent a format specifier that defines the required format of s.</param>
+        /// <param name="provider">An object that supplies culture-specific formatting information about s.</param>
+        /// <param name="style">A bitwise combination of enumeration values that indicates the permitted format of s. A typical value to specify is None.</param>
+        /// <returns>An object that is equivalent to the date contained in s, as specified by format, provider, and style.</returns>
+        public static DateOnly ParseExact(ReadOnlySpan<char> s, ReadOnlySpan<char> format, IFormatProvider? provider, DateTimeStyles style = DateTimeStyles.None)
+        {
+            if ((style & ~DateTimeStyles.AllowWhiteSpaces) != 0)
+            {
+                throw new ArgumentException(SR.Argument_InvalidDateStyles, nameof(style));
+            }
+
+            if (format.Length == 1)
+            {
+                if (format[0] == 'o' || format[0] == 'O')
+                {
+                    format = OFormat;
+                    provider = CultureInfo.InvariantCulture.DateTimeFormat;
+                }
+                else if (format[0] == 'r' || format[0] == 'R')
+                {
+                    format = RFormat;
+                    provider = CultureInfo.InvariantCulture.DateTimeFormat;
+                }
+            }
+
+            DateTimeResult result = default;
+            result.Init(s);
+
+            if (!DateTimeParse.TryParseExact(s, format, DateTimeFormatInfo.GetInstance(provider), style, ref result))
+            {
+                throw new FormatException(SR.Format(SR.Format_BadDateOnly, s.ToString()));
+            }
+
+            if ((result.flags & ParseFlagsDateMask) != 0)
+            {
+                throw new FormatException(SR.Format(SR.Format_DateOnlyContainsNoneDateParts, s.ToString()));
+            }
+
+            return new DateOnly(DayNumberFromDateTime(result.parsedDate));
+        }
```

Comment:
There's a lot of duplication between the Try and non-Try variants.  Can the non-Try variants just wrap the Try ones and throw an exception for failure?  It's what DateTime does, e.g.
```C#
        internal static DateTime Parse(ReadOnlySpan<char> s, DateTimeFormatInfo dtfi, DateTimeStyles styles)
        {
            DateTimeResult result = default;
            result.Init(s);

            if (!TryParse(s, dtfi, styles, ref result))
            {
                throw GetDateTimeParseException(ref result);
            }
            
            return result.parsedDate;
        }
```

Reply:
I was keeping it separate because we throw different exceptions. I'll look if we can consolidate these.

Reply:
I'll remove all duplications